json.go

```diff
@@ -0,0 +1,314 @@
+/*
+The purpose of the JSON marshalling and unmarshalling code is to
+ensure that any interface's concrete types are encoded with disambiguation
+in order to extract out the concrete type when decoding
+For example given
+type Transport struct {
+    Vehicle
+    Capacity int
+}
+
+type Vehicle interface {
+    Move() error
+}
+
+type Car   string
+type Boat  int
+type Plane int
+
+func (c Car) Move() error { return nil }
+func (b Boat) Move() error { return nil }
+func (p Plane) Move() error { return nil }
+
+wire.RegisterConcrete(&Transport{}, "our/transport", nil)
+wire.RegisterConcrete(&Car("Acura"), "car", nil)
+wire.MarshalJSON(&Transport{Car{}, 10})
+should give
+{
+  "_df": "XXXXXXXXXXXXXX",
+  "Vehicle":  "Acura",
+  "Capacity": 10
+}
+
+which will strictly only unmarshal to:
+Transport{Car("Acura"), Capacity: 10}
+given:
+
+tr := new(Transport)
+wire.UnmarshalJSON(blob, tr)
+*/
+package wire
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+	"unicode"
+)
+
+func (cdc *Codec) MarshalJSON(o interface{}) ([]byte, error) {
+	// reflect.ValueOf retrieves the concrete type
+	// that the interface contains.
+	val := reflect.ValueOf(o)
+	origVal := val
+
+	switch val.Kind() {
+	case reflect.Invalid: // They passed in: an untyped nil, marshal it to "null"
+		return []byte("null"), nil
+	case reflect.Ptr:
+		if val.IsNil() {
+			return []byte("null"), nil
+		}
+
+		// Perfect we've got a pointer
+		// Dereference it to the non-pointer type
+		val = deref(val)
+	}
+
+	// If the type implements json.MarshalJSON, then defer to its MarshalJSON method.
+	if anyImplementsJSONMarshaler(val, origVal) {
+		return json.Marshal(o)
+	}
+
+	// Except for:
+	//  * maps
+	//  * structs
+	//  * arrays and slices
+	// all other types should be encoded as they are.
+	switch val.Kind() {
+	case reflect.Map:
```

Comment:
We shouldn't support maps... the goal is compatibility between binary and json.. and we've never needed to support them.

Reply:
Hmm, I didn't realize that. Perhaps let's also document such if not already documented.